File: array_and_strings/sliding_window/python3/k_radius_avg.py
from typing import List

def find_average(nums: List[int], k: int) -> List[int]:
    """
    The k-radius average for a subarray of `nums` centered at index `i` with the radius `k` is the average of all elements in `nums` between the 
    indices `i - k` and `i + k` (inclusive). If there are less than `k` elements before or after the index `i`, then the k-radius average is `-1`.
    Build and return an array `avgs` of length `n` where `avgs[i]` is the k-radius average for the subarray centered at index `i`.

    Parameters
    ----------
    nums : List[int]
        A list of integers
    k : int
        A non-negative integer

    Returns
    -------
    List[int]
        The list of k-radius averages for the subarrays centered at each index of `nums`
    """
    window_size = 2 * k + 1
    # Prefill with -1 costing O(n)
    avg = [-1] * len(nums)
    
    # In this case, no element has k elements before or after it
    if (window_size > len(nums)):
        return avg
    # If radius is 0, the k-radius average for each element is simply itself
    if (k == 0):
        return nums
    
    # Find first window sum costing O(window_size)
    current_window_sum = sum(nums[slice(0, window_size)])
    # The first index with enough elements before it is k, and the first average is associated with that element
    avg[k] = current_window_sum // window_size
    
    # Only iterate through elements that have k elements before or after it
    for i in range(k + 1, len(nums) - k):
        # Current window bounds
        left = i - k 
        right = i + k
        # Slide the window by adding the next element 'nums[right]'
        # Keep window fixed by removing the leftmost element (i.e. nums[left - 1])
        current_window_sum += (nums[right] - nums[left - 1])
        avg[i] = current_window_sum // window_size
            
    return avg
    
def main() -> int:
    
    nums = [7, 4, 3, 9, 1, 8, 5, 2, 6]
    k = 3
    avg = find_average(nums, k)
    print(f"Give the input {nums} and k = {k}, the k-radius average is {avg}")
    
    return 0

if __name__ == "__main__":
    
    main()

File: array_and_strings/sliding_window/c++/k_radius_avg/k_radius_avg.cpp
// k_radius_avg.cpp
#include "k_radius_avg.hpp"
#include <vector>
#include <numeric> // For std::accumulate

std::vector<int> findAverage(const std::vector<int> &nums, int k)
{
    int windowSize = 2 * k + 1;
    // Prefill with -1
    std::vector<int> avg(nums.size(), -1);

    // In this case, no element has k elements before or after it
    if (windowSize > nums.size())
    {
        return avg;
    }
    // If k == 0, then the k-radius average for each element is itself
    if (k == 0)
    {
        return nums;
    }

    // First window costs O(windowSize)
    // Accumulate sums from [first, last) with a initial sum of 0LL
    long long firstWindowSum = std::accumulate(nums.begin(), nums.begin() + windowSize, 0LL);
    // The first index with enough elements before it is 'k', and so the first non -1 k-radius average is associated with that
    avg[k] = firstWindowSum / windowSize;

    // Only iterate through elements with valid window (i.e. k elements before or after)
    long long currentWindowSum = firstWindowSum;
    for (int i = k + 1; i < nums.size() - k; i++)
    {
        // These are the current window bounds
        int left = i - k;
        int right = i + k;
        // Expand the window by adding the next element pointed to by 'end'
        // Shrink the window (to keep it fixed) by removing the leftmost element pointed to by 'start - 1'
        currentWindowSum += (nums[right] - nums[left - 1]);
        avg[i] = currentWindowSum / windowSize;
    }

    return avg;
}
